// smscenter/SmsCenter.js
//
// Autogenerated by Thrift Compiler (0.9.2)
//
// DO NOT EDIT UNLESS YOU ARE SURE THAT YOU KNOW WHAT YOU ARE DOING
//
var thrift = require('thrift');
var Thrift = thrift.Thrift;
var Q = thrift.Q;


var ttypes = require('./smscenter_types');
//HELPER FUNCTIONS AND STRUCTURES

ThriftSmsCenter.SmsCenter_sendSms_args = function(args) {
  this.message = null;
  this.recipients = null;
  if (args) {
    if (args.message !== undefined) {
      this.message = args.message;
    }
    if (args.recipients !== undefined) {
      this.recipients = args.recipients;
    }
  }
};
ThriftSmsCenter.SmsCenter_sendSms_args.prototype = {};
ThriftSmsCenter.SmsCenter_sendSms_args.prototype.read = function(input) {
  input.readStructBegin();
  while (true)
  {
    var ret = input.readFieldBegin();
    var fname = ret.fname;
    var ftype = ret.ftype;
    var fid = ret.fid;
    if (ftype == Thrift.Type.STOP) {
      break;
    }
    switch (fid)
    {
      case 1:
      if (ftype == Thrift.Type.STRING) {
        this.message = input.readString();
      } else {
        input.skip(ftype);
      }
      break;
      case 2:
      if (ftype == Thrift.Type.LIST) {
        var _size0 = 0;
        var _rtmp34;
        this.recipients = [];
        var _etype3 = 0;
        _rtmp34 = input.readListBegin();
        _etype3 = _rtmp34.etype;
        _size0 = _rtmp34.size;
        for (var _i5 = 0; _i5 < _size0; ++_i5)
        {
          var elem6 = null;
          elem6 = input.readString();
          this.recipients.push(elem6);
        }
        input.readListEnd();
      } else {
        input.skip(ftype);
      }
      break;
      default:
        input.skip(ftype);
    }
    input.readFieldEnd();
  }
  input.readStructEnd();
  return;
};

ThriftSmsCenter.SmsCenter_sendSms_args.prototype.write = function(output) {
  output.writeStructBegin('SmsCenter_sendSms_args');
  if (this.message !== null && this.message !== undefined) {
    output.writeFieldBegin('message', Thrift.Type.STRING, 1);
    output.writeString(this.message);
    output.writeFieldEnd();
  }
  if (this.recipients !== null && this.recipients !== undefined) {
    output.writeFieldBegin('recipients', Thrift.Type.LIST, 2);
    output.writeListBegin(Thrift.Type.STRING, this.recipients.length);
    for (var iter7 in this.recipients)
    {
      if (this.recipients.hasOwnProperty(iter7))
      {
        iter7 = this.recipients[iter7];
        output.writeString(iter7);
      }
    }
    output.writeListEnd();
    output.writeFieldEnd();
  }
  output.writeFieldStop();
  output.writeStructEnd();
  return;
};

ThriftSmsCenter.SmsCenter_sendSms_result = function(args) {
  this.success = null;
  this.ex = null;
  if (args instanceof ttypes.SmsCenterException) {
    this.ex = args;
    return;
  }
  if (args) {
    if (args.success !== undefined) {
      this.success = args.success;
    }
    if (args.ex !== undefined) {
      this.ex = args.ex;
    }
  }
};
ThriftSmsCenter.SmsCenter_sendSms_result.prototype = {};
ThriftSmsCenter.SmsCenter_sendSms_result.prototype.read = function(input) {
  input.readStructBegin();
  while (true)
  {
    var ret = input.readFieldBegin();
    var fname = ret.fname;
    var ftype = ret.ftype;
    var fid = ret.fid;
    if (ftype == Thrift.Type.STOP) {
      break;
    }
    switch (fid)
    {
      case 0:
      if (ftype == Thrift.Type.STRING) {
        this.success = input.readString();
      } else {
        input.skip(ftype);
      }
      break;
      case 1:
      if (ftype == Thrift.Type.STRUCT) {
        this.ex = new ttypes.SmsCenterException();
        this.ex.read(input);
      } else {
        input.skip(ftype);
      }
      break;
      default:
        input.skip(ftype);
    }
    input.readFieldEnd();
  }
  input.readStructEnd();
  return;
};

ThriftSmsCenter.SmsCenter_sendSms_result.prototype.write = function(output) {
  output.writeStructBegin('SmsCenter_sendSms_result');
  if (this.success !== null && this.success !== undefined) {
    output.writeFieldBegin('success', Thrift.Type.STRING, 0);
    output.writeString(this.success);
    output.writeFieldEnd();
  }
  if (this.ex !== null && this.ex !== undefined) {
    output.writeFieldBegin('ex', Thrift.Type.STRUCT, 1);
    this.ex.write(output);
    output.writeFieldEnd();
  }
  output.writeFieldStop();
  output.writeStructEnd();
  return;
};

ThriftSmsCenter.SmsCenter__ping_args = function(args) {
};
ThriftSmsCenter.SmsCenter__ping_args.prototype = {};
ThriftSmsCenter.SmsCenter__ping_args.prototype.read = function(input) {
  input.readStructBegin();
  while (true)
  {
    var ret = input.readFieldBegin();
    var fname = ret.fname;
    var ftype = ret.ftype;
    var fid = ret.fid;
    if (ftype == Thrift.Type.STOP) {
      break;
    }
    input.skip(ftype);
    input.readFieldEnd();
  }
  input.readStructEnd();
  return;
};

ThriftSmsCenter.SmsCenter__ping_args.prototype.write = function(output) {
  output.writeStructBegin('SmsCenter__ping_args');
  output.writeFieldStop();
  output.writeStructEnd();
  return;
};

ThriftSmsCenter.SmsCenter__ping_result = function(args) {
  this.success = null;
  if (args) {
    if (args.success !== undefined) {
      this.success = args.success;
    }
  }
};
ThriftSmsCenter.SmsCenter__ping_result.prototype = {};
ThriftSmsCenter.SmsCenter__ping_result.prototype.read = function(input) {
  input.readStructBegin();
  while (true)
  {
    var ret = input.readFieldBegin();
    var fname = ret.fname;
    var ftype = ret.ftype;
    var fid = ret.fid;
    if (ftype == Thrift.Type.STOP) {
      break;
    }
    switch (fid)
    {
      case 0:
      if (ftype == Thrift.Type.STRING) {
        this.success = input.readString();
      } else {
        input.skip(ftype);
      }
      break;
      case 0:
        input.skip(ftype);
        break;
      default:
        input.skip(ftype);
    }
    input.readFieldEnd();
  }
  input.readStructEnd();
  return;
};

ThriftSmsCenter.SmsCenter__ping_result.prototype.write = function(output) {
  output.writeStructBegin('SmsCenter__ping_result');
  if (this.success !== null && this.success !== undefined) {
    output.writeFieldBegin('success', Thrift.Type.STRING, 0);
    output.writeString(this.success);
    output.writeFieldEnd();
  }
  output.writeFieldStop();
  output.writeStructEnd();
  return;
};

ThriftSmsCenter.SmsCenterClient = exports.Client = function(output, pClass) {
    this.output = output;
    this.pClass = pClass;
    this._seqid = 0;
    this._reqs = {};
};
ThriftSmsCenter.SmsCenterClient.prototype = {};
ThriftSmsCenter.SmsCenterClient.prototype.seqid = function() { return this._seqid; }
ThriftSmsCenter.SmsCenterClient.prototype.new_seqid = function() { return this._seqid += 1; }
ThriftSmsCenter.SmsCenterClient.prototype.sendSms = function(message, recipients, callback) {
  this._seqid = this.new_seqid();
  if (callback === undefined) {
    var _defer = Q.defer();
    this._reqs[this.seqid()] = function(error, result) {
      if (error) {
        _defer.reject(error);
      } else {
        _defer.resolve(result);
      }
    };
    this.send_sendSms(message, recipients);
    return _defer.promise;
  } else {
    this._reqs[this.seqid()] = callback;
    this.send_sendSms(message, recipients);
  }
};

ThriftSmsCenter.SmsCenterClient.prototype.send_sendSms = function(message, recipients) {
  var output = new this.pClass(this.output);
  output.writeMessageBegin('sendSms', Thrift.MessageType.CALL, this.seqid());
  var args = new ThriftSmsCenter.SmsCenter_sendSms_args();
  args.message = message;
  args.recipients = recipients;
  args.write(output);
  output.writeMessageEnd();
  return this.output.flush();
};

ThriftSmsCenter.SmsCenterClient.prototype.recv_sendSms = function(input,mtype,rseqid) {
  var callback = this._reqs[rseqid] || function() {};
  delete this._reqs[rseqid];
  if (mtype == Thrift.MessageType.EXCEPTION) {
    var x = new Thrift.TApplicationException();
    x.read(input);
    input.readMessageEnd();
    return callback(x);
  }
  var result = new ThriftSmsCenter.SmsCenter_sendSms_result();
  result.read(input);
  input.readMessageEnd();

  if (null !== result.ex) {
    return callback(result.ex);
  }
  if (null !== result.success) {
    return callback(null, result.success);
  }
  return callback('sendSms failed: unknown result');
};
ThriftSmsCenter.SmsCenterClient.prototype._ping = function(callback) {
  this._seqid = this.new_seqid();
  if (callback === undefined) {
    var _defer = Q.defer();
    this._reqs[this.seqid()] = function(error, result) {
      if (error) {
        _defer.reject(error);
      } else {
        _defer.resolve(result);
      }
    };
    this.send__ping();
    return _defer.promise;
  } else {
    this._reqs[this.seqid()] = callback;
    this.send__ping();
  }
};

ThriftSmsCenter.SmsCenterClient.prototype.send__ping = function() {
  var output = new this.pClass(this.output);
  output.writeMessageBegin('_ping', Thrift.MessageType.CALL, this.seqid());
  var args = new ThriftSmsCenter.SmsCenter__ping_args();
  args.write(output);
  output.writeMessageEnd();
  return this.output.flush();
};

ThriftSmsCenter.SmsCenterClient.prototype.recv__ping = function(input,mtype,rseqid) {
  var callback = this._reqs[rseqid] || function() {};
  delete this._reqs[rseqid];
  if (mtype == Thrift.MessageType.EXCEPTION) {
    var x = new Thrift.TApplicationException();
    x.read(input);
    input.readMessageEnd();
    return callback(x);
  }
  var result = new ThriftSmsCenter.SmsCenter__ping_result();
  result.read(input);
  input.readMessageEnd();

  if (null !== result.success) {
    return callback(null, result.success);
  }
  return callback('_ping failed: unknown result');
};
ThriftSmsCenter.SmsCenterProcessor = exports.Processor = function(handler) {
  this._handler = handler
}
ThriftSmsCenter.SmsCenterProcessor.prototype.process = function(input, output) {
  var r = input.readMessageBegin();
  if (this['process_' + r.fname]) {
    return this['process_' + r.fname].call(this, r.rseqid, input, output);
  } else {
    input.skip(Thrift.Type.STRUCT);
    input.readMessageEnd();
    var x = new Thrift.TApplicationException(Thrift.TApplicationExceptionType.UNKNOWN_METHOD, 'Unknown function ' + r.fname);
    output.writeMessageBegin(r.fname, Thrift.MessageType.EXCEPTION, r.rseqid);
    x.write(output);
    output.writeMessageEnd();
    output.flush();
  }
}

ThriftSmsCenter.SmsCenterProcessor.prototype.process_sendSms = function(seqid, input, output) {
  var args = new ThriftSmsCenter.SmsCenter_sendSms_args();
  args.read(input);
  input.readMessageEnd();
  if (this._handler.sendSms.length === 2) {
    Q.fcall(this._handler.sendSms, args.message, args.recipients)
      .then(function(result) {
        var result = new ThriftSmsCenter.SmsCenter_sendSms_result({success: result});
        output.writeMessageBegin("sendSms", Thrift.MessageType.REPLY, seqid);
        result.write(output);
        output.writeMessageEnd();
        output.flush();
      }, function (err) {
        var result = new ThriftSmsCenter.SmsCenter_sendSms_result(err);
        output.writeMessageBegin("sendSms", Thrift.MessageType.REPLY, seqid);
        result.write(output);
        output.writeMessageEnd();
        output.flush();
      });
  } else {
    this._handler.sendSms(args.message, args.recipients,  function (err, result) {
      var result = new ThriftSmsCenter.SmsCenter_sendSms_result((err != null ? err : {success: result}));
      output.writeMessageBegin("sendSms", Thrift.MessageType.REPLY, seqid);
      result.write(output);
      output.writeMessageEnd();
      output.flush();
    });
  }
}

ThriftSmsCenter.SmsCenterProcessor.prototype.process__ping = function(seqid, input, output) {
  var args = new ThriftSmsCenter.SmsCenter__ping_args();
  args.read(input);
  input.readMessageEnd();
  if (this._handler._ping.length === 0) {
    Q.fcall(this._handler._ping)
      .then(function(result) {
        var result = new ThriftSmsCenter.SmsCenter__ping_result({success: result});
        output.writeMessageBegin("_ping", Thrift.MessageType.REPLY, seqid);
        result.write(output);
        output.writeMessageEnd();
        output.flush();
      }, function (err) {
        var result = new ThriftSmsCenter.SmsCenter__ping_result(err);
        output.writeMessageBegin("_ping", Thrift.MessageType.REPLY, seqid);
        result.write(output);
        output.writeMessageEnd();
        output.flush();
      });
  } else {
    this._handler._ping( function (err, result) {
      var result = new ThriftSmsCenter.SmsCenter__ping_result((err != null ? err : {success: result}));
      output.writeMessageBegin("_ping", Thrift.MessageType.REPLY, seqid);
      result.write(output);
      output.writeMessageEnd();
      output.flush();
    });
  }
}
