// idgen/IdGen.js
//
// Autogenerated by Thrift Compiler (0.9.3)
//
// DO NOT EDIT UNLESS YOU ARE SURE THAT YOU KNOW WHAT YOU ARE DOING
//
var thrift = require('thrift');
var Thrift = thrift.Thrift;
var Q = thrift.Q;


var ttypes = require('./service_types');
//HELPER FUNCTIONS AND STRUCTURES

IdGen_ping_args = function(args) {
};
IdGen_ping_args.prototype = {};
IdGen_ping_args.prototype.read = function(input) {
  input.readStructBegin();
  while (true)
  {
    var ret = input.readFieldBegin();
    var fname = ret.fname;
    var ftype = ret.ftype;
    var fid = ret.fid;
    if (ftype == Thrift.Type.STOP) {
      break;
    }
    input.skip(ftype);
    input.readFieldEnd();
  }
  input.readStructEnd();
  return;
};

IdGen_ping_args.prototype.write = function(output) {
  output.writeStructBegin('IdGen_ping_args');
  output.writeFieldStop();
  output.writeStructEnd();
  return;
};

IdGen_ping_result = function(args) {
  this.success = null;
  if (args) {
    if (args.success !== undefined && args.success !== null) {
      this.success = args.success;
    }
  }
};
IdGen_ping_result.prototype = {};
IdGen_ping_result.prototype.read = function(input) {
  input.readStructBegin();
  while (true)
  {
    var ret = input.readFieldBegin();
    var fname = ret.fname;
    var ftype = ret.ftype;
    var fid = ret.fid;
    if (ftype == Thrift.Type.STOP) {
      break;
    }
    switch (fid)
    {
      case 0:
      if (ftype == Thrift.Type.STRING) {
        this.success = input.readString();
      } else {
        input.skip(ftype);
      }
      break;
      case 0:
        input.skip(ftype);
        break;
      default:
        input.skip(ftype);
    }
    input.readFieldEnd();
  }
  input.readStructEnd();
  return;
};

IdGen_ping_result.prototype.write = function(output) {
  output.writeStructBegin('IdGen_ping_result');
  if (this.success !== null && this.success !== undefined) {
    output.writeFieldBegin('success', Thrift.Type.STRING, 0);
    output.writeString(this.success);
    output.writeFieldEnd();
  }
  output.writeFieldStop();
  output.writeStructEnd();
  return;
};

IdGen_generate_args = function(args) {
  this.generator = null;
  if (args) {
    if (args.generator !== undefined && args.generator !== null) {
      this.generator = args.generator;
    }
  }
};
IdGen_generate_args.prototype = {};
IdGen_generate_args.prototype.read = function(input) {
  input.readStructBegin();
  while (true)
  {
    var ret = input.readFieldBegin();
    var fname = ret.fname;
    var ftype = ret.ftype;
    var fid = ret.fid;
    if (ftype == Thrift.Type.STOP) {
      break;
    }
    switch (fid)
    {
      case 1:
      if (ftype == Thrift.Type.STRING) {
        this.generator = input.readString();
      } else {
        input.skip(ftype);
      }
      break;
      case 0:
        input.skip(ftype);
        break;
      default:
        input.skip(ftype);
    }
    input.readFieldEnd();
  }
  input.readStructEnd();
  return;
};

IdGen_generate_args.prototype.write = function(output) {
  output.writeStructBegin('IdGen_generate_args');
  if (this.generator !== null && this.generator !== undefined) {
    output.writeFieldBegin('generator', Thrift.Type.STRING, 1);
    output.writeString(this.generator);
    output.writeFieldEnd();
  }
  output.writeFieldStop();
  output.writeStructEnd();
  return;
};

IdGen_generate_result = function(args) {
  this.success = null;
  if (args) {
    if (args.success !== undefined && args.success !== null) {
      this.success = args.success;
    }
  }
};
IdGen_generate_result.prototype = {};
IdGen_generate_result.prototype.read = function(input) {
  input.readStructBegin();
  while (true)
  {
    var ret = input.readFieldBegin();
    var fname = ret.fname;
    var ftype = ret.ftype;
    var fid = ret.fid;
    if (ftype == Thrift.Type.STOP) {
      break;
    }
    switch (fid)
    {
      case 0:
      if (ftype == Thrift.Type.I64) {
        this.success = input.readI64();
      } else {
        input.skip(ftype);
      }
      break;
      case 0:
        input.skip(ftype);
        break;
      default:
        input.skip(ftype);
    }
    input.readFieldEnd();
  }
  input.readStructEnd();
  return;
};

IdGen_generate_result.prototype.write = function(output) {
  output.writeStructBegin('IdGen_generate_result');
  if (this.success !== null && this.success !== undefined) {
    output.writeFieldBegin('success', Thrift.Type.I64, 0);
    output.writeI64(this.success);
    output.writeFieldEnd();
  }
  output.writeFieldStop();
  output.writeStructEnd();
  return;
};

IdGen_getCounter_args = function(args) {
  this.generator = null;
  if (args) {
    if (args.generator !== undefined && args.generator !== null) {
      this.generator = args.generator;
    }
  }
};
IdGen_getCounter_args.prototype = {};
IdGen_getCounter_args.prototype.read = function(input) {
  input.readStructBegin();
  while (true)
  {
    var ret = input.readFieldBegin();
    var fname = ret.fname;
    var ftype = ret.ftype;
    var fid = ret.fid;
    if (ftype == Thrift.Type.STOP) {
      break;
    }
    switch (fid)
    {
      case 1:
      if (ftype == Thrift.Type.STRING) {
        this.generator = input.readString();
      } else {
        input.skip(ftype);
      }
      break;
      case 0:
        input.skip(ftype);
        break;
      default:
        input.skip(ftype);
    }
    input.readFieldEnd();
  }
  input.readStructEnd();
  return;
};

IdGen_getCounter_args.prototype.write = function(output) {
  output.writeStructBegin('IdGen_getCounter_args');
  if (this.generator !== null && this.generator !== undefined) {
    output.writeFieldBegin('generator', Thrift.Type.STRING, 1);
    output.writeString(this.generator);
    output.writeFieldEnd();
  }
  output.writeFieldStop();
  output.writeStructEnd();
  return;
};

IdGen_getCounter_result = function(args) {
  this.success = null;
  if (args) {
    if (args.success !== undefined && args.success !== null) {
      this.success = args.success;
    }
  }
};
IdGen_getCounter_result.prototype = {};
IdGen_getCounter_result.prototype.read = function(input) {
  input.readStructBegin();
  while (true)
  {
    var ret = input.readFieldBegin();
    var fname = ret.fname;
    var ftype = ret.ftype;
    var fid = ret.fid;
    if (ftype == Thrift.Type.STOP) {
      break;
    }
    switch (fid)
    {
      case 0:
      if (ftype == Thrift.Type.I64) {
        this.success = input.readI64();
      } else {
        input.skip(ftype);
      }
      break;
      case 0:
        input.skip(ftype);
        break;
      default:
        input.skip(ftype);
    }
    input.readFieldEnd();
  }
  input.readStructEnd();
  return;
};

IdGen_getCounter_result.prototype.write = function(output) {
  output.writeStructBegin('IdGen_getCounter_result');
  if (this.success !== null && this.success !== undefined) {
    output.writeFieldBegin('success', Thrift.Type.I64, 0);
    output.writeI64(this.success);
    output.writeFieldEnd();
  }
  output.writeFieldStop();
  output.writeStructEnd();
  return;
};

IdGen_resetCounter_args = function(args) {
  this.generator = null;
  this.level = null;
  if (args) {
    if (args.generator !== undefined && args.generator !== null) {
      this.generator = args.generator;
    }
    if (args.level !== undefined && args.level !== null) {
      this.level = args.level;
    }
  }
};
IdGen_resetCounter_args.prototype = {};
IdGen_resetCounter_args.prototype.read = function(input) {
  input.readStructBegin();
  while (true)
  {
    var ret = input.readFieldBegin();
    var fname = ret.fname;
    var ftype = ret.ftype;
    var fid = ret.fid;
    if (ftype == Thrift.Type.STOP) {
      break;
    }
    switch (fid)
    {
      case 1:
      if (ftype == Thrift.Type.STRING) {
        this.generator = input.readString();
      } else {
        input.skip(ftype);
      }
      break;
      case 2:
      if (ftype == Thrift.Type.I64) {
        this.level = input.readI64();
      } else {
        input.skip(ftype);
      }
      break;
      default:
        input.skip(ftype);
    }
    input.readFieldEnd();
  }
  input.readStructEnd();
  return;
};

IdGen_resetCounter_args.prototype.write = function(output) {
  output.writeStructBegin('IdGen_resetCounter_args');
  if (this.generator !== null && this.generator !== undefined) {
    output.writeFieldBegin('generator', Thrift.Type.STRING, 1);
    output.writeString(this.generator);
    output.writeFieldEnd();
  }
  if (this.level !== null && this.level !== undefined) {
    output.writeFieldBegin('level', Thrift.Type.I64, 2);
    output.writeI64(this.level);
    output.writeFieldEnd();
  }
  output.writeFieldStop();
  output.writeStructEnd();
  return;
};

IdGen_resetCounter_result = function(args) {
};
IdGen_resetCounter_result.prototype = {};
IdGen_resetCounter_result.prototype.read = function(input) {
  input.readStructBegin();
  while (true)
  {
    var ret = input.readFieldBegin();
    var fname = ret.fname;
    var ftype = ret.ftype;
    var fid = ret.fid;
    if (ftype == Thrift.Type.STOP) {
      break;
    }
    input.skip(ftype);
    input.readFieldEnd();
  }
  input.readStructEnd();
  return;
};

IdGen_resetCounter_result.prototype.write = function(output) {
  output.writeStructBegin('IdGen_resetCounter_result');
  output.writeFieldStop();
  output.writeStructEnd();
  return;
};

IdGenClient = exports.Client = function(output, pClass) {
    this.output = output;
    this.pClass = pClass;
    this._seqid = 0;
    this._reqs = {};
};
IdGenClient.prototype = {};
IdGenClient.prototype.seqid = function() { return this._seqid; }
IdGenClient.prototype.new_seqid = function() { return this._seqid += 1; }
IdGenClient.prototype.ping = function(callback) {
  this._seqid = this.new_seqid();
  if (callback === undefined) {
    var _defer = Q.defer();
    this._reqs[this.seqid()] = function(error, result) {
      if (error) {
        _defer.reject(error);
      } else {
        _defer.resolve(result);
      }
    };
    this.send_ping();
    return _defer.promise;
  } else {
    this._reqs[this.seqid()] = callback;
    this.send_ping();
  }
};

IdGenClient.prototype.send_ping = function() {
  var output = new this.pClass(this.output);
  output.writeMessageBegin('ping', Thrift.MessageType.CALL, this.seqid());
  var args = new IdGen_ping_args();
  args.write(output);
  output.writeMessageEnd();
  return this.output.flush();
};

IdGenClient.prototype.recv_ping = function(input,mtype,rseqid) {
  var callback = this._reqs[rseqid] || function() {};
  delete this._reqs[rseqid];
  if (mtype == Thrift.MessageType.EXCEPTION) {
    var x = new Thrift.TApplicationException();
    x.read(input);
    input.readMessageEnd();
    return callback(x);
  }
  var result = new IdGen_ping_result();
  result.read(input);
  input.readMessageEnd();

  if (null !== result.success) {
    return callback(null, result.success);
  }
  return callback('ping failed: unknown result');
};
IdGenClient.prototype.generate = function(generator, callback) {
  this._seqid = this.new_seqid();
  if (callback === undefined) {
    var _defer = Q.defer();
    this._reqs[this.seqid()] = function(error, result) {
      if (error) {
        _defer.reject(error);
      } else {
        _defer.resolve(result);
      }
    };
    this.send_generate(generator);
    return _defer.promise;
  } else {
    this._reqs[this.seqid()] = callback;
    this.send_generate(generator);
  }
};

IdGenClient.prototype.send_generate = function(generator) {
  var output = new this.pClass(this.output);
  output.writeMessageBegin('generate', Thrift.MessageType.CALL, this.seqid());
  var args = new IdGen_generate_args();
  args.generator = generator;
  args.write(output);
  output.writeMessageEnd();
  return this.output.flush();
};

IdGenClient.prototype.recv_generate = function(input,mtype,rseqid) {
  var callback = this._reqs[rseqid] || function() {};
  delete this._reqs[rseqid];
  if (mtype == Thrift.MessageType.EXCEPTION) {
    var x = new Thrift.TApplicationException();
    x.read(input);
    input.readMessageEnd();
    return callback(x);
  }
  var result = new IdGen_generate_result();
  result.read(input);
  input.readMessageEnd();

  if (null !== result.success) {
    return callback(null, result.success);
  }
  return callback('generate failed: unknown result');
};
IdGenClient.prototype.getCounter = function(generator, callback) {
  this._seqid = this.new_seqid();
  if (callback === undefined) {
    var _defer = Q.defer();
    this._reqs[this.seqid()] = function(error, result) {
      if (error) {
        _defer.reject(error);
      } else {
        _defer.resolve(result);
      }
    };
    this.send_getCounter(generator);
    return _defer.promise;
  } else {
    this._reqs[this.seqid()] = callback;
    this.send_getCounter(generator);
  }
};

IdGenClient.prototype.send_getCounter = function(generator) {
  var output = new this.pClass(this.output);
  output.writeMessageBegin('getCounter', Thrift.MessageType.CALL, this.seqid());
  var args = new IdGen_getCounter_args();
  args.generator = generator;
  args.write(output);
  output.writeMessageEnd();
  return this.output.flush();
};

IdGenClient.prototype.recv_getCounter = function(input,mtype,rseqid) {
  var callback = this._reqs[rseqid] || function() {};
  delete this._reqs[rseqid];
  if (mtype == Thrift.MessageType.EXCEPTION) {
    var x = new Thrift.TApplicationException();
    x.read(input);
    input.readMessageEnd();
    return callback(x);
  }
  var result = new IdGen_getCounter_result();
  result.read(input);
  input.readMessageEnd();

  if (null !== result.success) {
    return callback(null, result.success);
  }
  return callback('getCounter failed: unknown result');
};
IdGenClient.prototype.resetCounter = function(generator, level, callback) {
  this._seqid = this.new_seqid();
  if (callback === undefined) {
    var _defer = Q.defer();
    this._reqs[this.seqid()] = function(error, result) {
      if (error) {
        _defer.reject(error);
      } else {
        _defer.resolve(result);
      }
    };
    this.send_resetCounter(generator, level);
    return _defer.promise;
  } else {
    this._reqs[this.seqid()] = callback;
    this.send_resetCounter(generator, level);
  }
};

IdGenClient.prototype.send_resetCounter = function(generator, level) {
  var output = new this.pClass(this.output);
  output.writeMessageBegin('resetCounter', Thrift.MessageType.CALL, this.seqid());
  var args = new IdGen_resetCounter_args();
  args.generator = generator;
  args.level = level;
  args.write(output);
  output.writeMessageEnd();
  return this.output.flush();
};

IdGenClient.prototype.recv_resetCounter = function(input,mtype,rseqid) {
  var callback = this._reqs[rseqid] || function() {};
  delete this._reqs[rseqid];
  if (mtype == Thrift.MessageType.EXCEPTION) {
    var x = new Thrift.TApplicationException();
    x.read(input);
    input.readMessageEnd();
    return callback(x);
  }
  var result = new IdGen_resetCounter_result();
  result.read(input);
  input.readMessageEnd();

  callback(null)
};
IdGenProcessor = exports.Processor = function(handler) {
  this._handler = handler
}
IdGenProcessor.prototype.process = function(input, output) {
  var r = input.readMessageBegin();
  if (this['process_' + r.fname]) {
    return this['process_' + r.fname].call(this, r.rseqid, input, output);
  } else {
    input.skip(Thrift.Type.STRUCT);
    input.readMessageEnd();
    var x = new Thrift.TApplicationException(Thrift.TApplicationExceptionType.UNKNOWN_METHOD, 'Unknown function ' + r.fname);
    output.writeMessageBegin(r.fname, Thrift.MessageType.EXCEPTION, r.rseqid);
    x.write(output);
    output.writeMessageEnd();
    output.flush();
  }
}

IdGenProcessor.prototype.process_ping = function(seqid, input, output) {
  var args = new IdGen_ping_args();
  args.read(input);
  input.readMessageEnd();
  if (this._handler.ping.length === 0) {
    Q.fcall(this._handler.ping)
      .then(function(result) {
        var result = new IdGen_ping_result({success: result});
        output.writeMessageBegin("ping", Thrift.MessageType.REPLY, seqid);
        result.write(output);
        output.writeMessageEnd();
        output.flush();
      }, function (err) {
        var result = new Thrift.TApplicationException(Thrift.TApplicationExceptionType.UNKNOWN, err.message);
        output.writeMessageBegin("ping", Thrift.MessageType.EXCEPTION, seqid);
        result.write(output);
        output.writeMessageEnd();
        output.flush();
      });
  } else {
    this._handler.ping(function (err, result) {
      if (err == null) {
        var result = new IdGen_ping_result((err != null ? err : {success: result}));
        output.writeMessageBegin("ping", Thrift.MessageType.REPLY, seqid);
      } else {
        var result = new Thrift.TApplicationException(Thrift.TApplicationExceptionType.UNKNOWN, err.message);
        output.writeMessageBegin("ping", Thrift.MessageType.EXCEPTION, seqid);
      }
      result.write(output);
      output.writeMessageEnd();
      output.flush();
    });
  }
}

IdGenProcessor.prototype.process_generate = function(seqid, input, output) {
  var args = new IdGen_generate_args();
  args.read(input);
  input.readMessageEnd();
  if (this._handler.generate.length === 1) {
    Q.fcall(this._handler.generate, args.generator)
      .then(function(result) {
        var result = new IdGen_generate_result({success: result});
        output.writeMessageBegin("generate", Thrift.MessageType.REPLY, seqid);
        result.write(output);
        output.writeMessageEnd();
        output.flush();
      }, function (err) {
        var result = new Thrift.TApplicationException(Thrift.TApplicationExceptionType.UNKNOWN, err.message);
        output.writeMessageBegin("generate", Thrift.MessageType.EXCEPTION, seqid);
        result.write(output);
        output.writeMessageEnd();
        output.flush();
      });
  } else {
    this._handler.generate(args.generator, function (err, result) {
      if (err == null) {
        var result = new IdGen_generate_result((err != null ? err : {success: result}));
        output.writeMessageBegin("generate", Thrift.MessageType.REPLY, seqid);
      } else {
        var result = new Thrift.TApplicationException(Thrift.TApplicationExceptionType.UNKNOWN, err.message);
        output.writeMessageBegin("generate", Thrift.MessageType.EXCEPTION, seqid);
      }
      result.write(output);
      output.writeMessageEnd();
      output.flush();
    });
  }
}

IdGenProcessor.prototype.process_getCounter = function(seqid, input, output) {
  var args = new IdGen_getCounter_args();
  args.read(input);
  input.readMessageEnd();
  if (this._handler.getCounter.length === 1) {
    Q.fcall(this._handler.getCounter, args.generator)
      .then(function(result) {
        var result = new IdGen_getCounter_result({success: result});
        output.writeMessageBegin("getCounter", Thrift.MessageType.REPLY, seqid);
        result.write(output);
        output.writeMessageEnd();
        output.flush();
      }, function (err) {
        var result = new Thrift.TApplicationException(Thrift.TApplicationExceptionType.UNKNOWN, err.message);
        output.writeMessageBegin("getCounter", Thrift.MessageType.EXCEPTION, seqid);
        result.write(output);
        output.writeMessageEnd();
        output.flush();
      });
  } else {
    this._handler.getCounter(args.generator, function (err, result) {
      if (err == null) {
        var result = new IdGen_getCounter_result((err != null ? err : {success: result}));
        output.writeMessageBegin("getCounter", Thrift.MessageType.REPLY, seqid);
      } else {
        var result = new Thrift.TApplicationException(Thrift.TApplicationExceptionType.UNKNOWN, err.message);
        output.writeMessageBegin("getCounter", Thrift.MessageType.EXCEPTION, seqid);
      }
      result.write(output);
      output.writeMessageEnd();
      output.flush();
    });
  }
}

IdGenProcessor.prototype.process_resetCounter = function(seqid, input, output) {
  var args = new IdGen_resetCounter_args();
  args.read(input);
  input.readMessageEnd();
  if (this._handler.resetCounter.length === 2) {
    Q.fcall(this._handler.resetCounter, args.generator, args.level)
      .then(function(result) {
        var result = new IdGen_resetCounter_result({success: result});
        output.writeMessageBegin("resetCounter", Thrift.MessageType.REPLY, seqid);
        result.write(output);
        output.writeMessageEnd();
        output.flush();
      }, function (err) {
        var result = new Thrift.TApplicationException(Thrift.TApplicationExceptionType.UNKNOWN, err.message);
        output.writeMessageBegin("resetCounter", Thrift.MessageType.EXCEPTION, seqid);
        result.write(output);
        output.writeMessageEnd();
        output.flush();
      });
  } else {
    this._handler.resetCounter(args.generator, args.level, function (err, result) {
      if (err == null) {
        var result = new IdGen_resetCounter_result((err != null ? err : {success: result}));
        output.writeMessageBegin("resetCounter", Thrift.MessageType.REPLY, seqid);
      } else {
        var result = new Thrift.TApplicationException(Thrift.TApplicationExceptionType.UNKNOWN, err.message);
        output.writeMessageBegin("resetCounter", Thrift.MessageType.EXCEPTION, seqid);
      }
      result.write(output);
      output.writeMessageEnd();
      output.flush();
    });
  }
}
